pl_env.c

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2021 Patrick Gaskin
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "pl_env.h"
+#include "debug.h"
+#include "options.h"
+#include "xmalloc.h"
+
+#include <ctype.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * pl_getenv_normalized gets an env var and puts it into the required format for
+ * pl_env_save and pl_env_restore, which does exact string matching/replacement
+ * against the file paths.
+ */
+static char *pl_env_getenv_normalized(const char *var)
```

Comment:
Even though this is currently run at least twice for each track on startup and shutdown, it shouldn't cause performance issues. I did some benchmarks, and it's only around 1-3 microseconds per run even for complex paths on a few of my devices, compiled with and without optimizations. I can reduce it to around 0.2 if I stack allocate the variables instead, but since the amount is negligible, at least for my library size, I'm going to leave it this way for now. I might implement caching later, though.

Reply:
Here's what it would look like (it still needs a bit of cleanup):

<details>

```diff
diff --git a/options.c b/options.c
index 627cc1f..9b1f4b3 100644
--- a/options.c
+++ b/options.c
@@ -41,6 +41,7 @@
 #include "debug.h"
 #include "discid.h"
 #include "mpris.h"
+#include "pl_env.h"
 
 #include <stdio.h>
 #include <errno.h>
diff --git a/pl_env.c b/pl_env.c
index 642401f..4eccc0e 100644
--- a/pl_env.c
+++ b/pl_env.c
@@ -16,7 +16,6 @@
  */
 
 #include "pl_env.h"
-#include "debug.h"
 #include "options.h"
 #include "xmalloc.h"
 
@@ -26,33 +25,14 @@
 #include <stdlib.h>
 #include <string.h>
 
+extern char **environ;
+
 /**
- * pl_getenv_normalized gets an env var and puts it into the required format for
- * pl_env_save and pl_env_restore, which does exact string matching/replacement
- * against the file paths.
+ * pl_env_norm normalizes the provided path into the required format for
+ * pl_env_save and pl_env_restore, which does exact string matching/replacement.
  */
-static char *pl_env_getenv_normalized(const char *var)
+static char *pl_env_norm(char *path)
 {
-       /* duplicate the var */
-       char *var_d = xstrdup(var);
-
-       /* trim the var */
-       char *var_s = var_d;
-       char *var_e = var_d + strlen(var_d);
-       while (var_s < var_e && isspace(*var_s)) var_s++;
-       while (var_e > var_s && isspace(*(var_e-1))) var_e--;
-       *var_e = '\0';
-
-       /* get the var */
-       char *val = getenv(var_s);
-       free(var_d);
-
-       /* if it's nonexistent or empty, return NULL */
-       if (!val || !*val) return NULL;
-
-       /* duplicate the value */
-       char *new = xstrdup(val);
-
 #ifdef _WIN32
        /* convert backslashes to slashes */
        /* note: cmus uses forward slashes internally, but Windows accepts both */
@@ -62,58 +42,58 @@ static char *pl_env_getenv_normalized(const char *var)
 
        /* canonicalize the path in-place */
        size_t r = 1, w = 1;
-       while (new[r]) {
+       while (path[r]) {
 
                /* handle the start of a segment */
-               if (w == 0 || new[w-1] == '/') {
+               if (w == 0 || path[w-1] == '/') {
 
                        /* handle empty segments */
-                       if (new[r] == '/') {
+                       if (path[r] == '/') {
 
                                /* skip the duplicate slashes */
-                               while (new[r] == '/') r++;
+                               while (path[r] == '/') r++;
 
                                continue;
                        }
 
                        /* handle '.' segments */
-                       if (new[r] == '.' && (new[r+1] == '/' || !new[r+1])) {
+                       if (path[r] == '.' && (path[r+1] == '/' || !path[r+1])) {
 
                                /* skip them */
-                               if (new[r += 1]) r++;
+                               if (path[r += 1]) r++;
 
                                continue;
                        }
 
                        /* handle '..' segments */
-                       if (new[r] == '.' && new[r+1] == '.' && (new[r+2] == '/' || !new[r+2])) {
+                       if (path[r] == '.' && path[r+1] == '.' && (path[r+2] == '/' || !path[r+2])) {
 
                                /* if there aren't any parent directories left to skip, return NULL */
                                if (!w) return NULL;
 
                                /* remove the previous segment up to the '/' */
-                               for (w--; w && new[w-1] != '/'; ) w--;
+                               for (w--; w && path[w-1] != '/'; ) w--;
 
                                /* skip the '..' */
-                               if (new[r += 2]) r++;
+                               if (path[r += 2]) r++;
 
                                continue;
                        }
                }
 
                /* write the next character */
-               new[w++] = new[r++];
+               path[w++] = path[r++];
        }
 
        /* remove the trailing slash if the path isn't / */
-       if (w >= 2 && new[w-1] == '/') {
+       if (w >= 2 && path[w-1] == '/') {
                w--;
        }
 
        /* terminate the path */
-       new[w] = '\0';
+       path[w] = '\0';
 
-       return new;
+       return path;
 }
 
 static bool pl_env_contains_marker(const char *str)
@@ -121,32 +101,67 @@ static bool pl_env_contains_marker(const char *str)
        return !!strchr(str, PL_ENV_DELIMITER);
 }
 
+const char *pl_env_cache(const char *var, int var_len)
+{
+       static char **cache = NULL;
+
+       if (!var) {
+               for (char **x = cache; x && *x; x++)
+                       free(*x);
+               free(cache);
+
+               size_t n = 0;
+               for (char **x = environ; *x; x++)
+                       n++;
+
+               char **new = cache = xnew(char*, n+1);
+               for (char **x = environ; *x; x++)
+                       if (!pl_env_contains_marker(*x))
+                               if (!pl_env_norm(strchr((*new++ = xstrdup(*x)), '=') + 1))
+                                       free(*--new);
+               *new = NULL;
+
+               return NULL;
+       }
+
+       size_t vl = var_len == -1
+               ? strlen(var)
+               : var_len;
+
+       const char *vs = var;
+       const char *ve = var + vl;
+       while (vs < ve && isspace(*vs)) vs++;
+       while (ve > vs && isspace(*(ve-1))) ve--;
+       vl = ve-vs;
+
+       if (memchr(vs, (int) vl, PL_ENV_DELIMITER) || memchr(vs, (int) vl, '='))
+               return NULL;
+
+       for (char **x = cache; x && *x; x++)
+               if (strncmp(*x, vs, vl) == 0 && (*x)[vl] == '=')
+                       return *x + vl + 1;
+
+       return NULL;
+}
+
 char *pl_env_save(const char *path)
 {
        if (!pl_env_vars || !*pl_env_vars || pl_env_var(path, NULL))
                return xstrdup(path);
        
        for (char **var = pl_env_vars; *var && **var; var++) {
-               char *val = pl_env_getenv_normalized(*var);
+               const char *val = pl_env_cache(*var, -1);
                if (!val)
                        continue;
-               if (!*val || pl_env_contains_marker(val)) {
-                       free(val);
-                       continue;
-               }
-               
+
                size_t val_len = strlen(val);
 
 #ifdef _WIN32
-               if (strncasecmp(path, val, val_len) != 0) {
-                       free(val);
+               if (strncasecmp(path, val, val_len) != 0)
                        continue;
-               }
 #else
-               if (strncmp(path, val, val_len) != 0) {
-                       free(val);
+               if (strncmp(path, val, val_len) != 0)
                        continue;
-               }
 #endif
 
                const char *rem = path + val_len;
@@ -154,10 +169,8 @@ char *pl_env_save(const char *path)
                /* always keep the slash at the beginning of the path, and only
                   use the env var if it replaces an entire path component (i.e.
                   it is a directory) */
-               if (*rem != '/') {
-                       free(val);
+               if (*rem != '/')
                        continue;
-               }
 
                size_t var_len = strlen(*var);
                size_t rem_len = strlen(rem);
@@ -172,7 +185,6 @@ char *pl_env_save(const char *path)
                ptr += rem_len;
                *ptr = '\0';
 
-               free(val);
                return new;
        }
 
@@ -184,20 +196,16 @@ char *pl_env_restore(const char *path)
        if (!path)
                return NULL;
 
-       const char *rem;
-       char *var = pl_env_var_str(path, &rem);
-       if (!var)
+       int len;
+       const char *var;
+       if (!(var = pl_env_var(path, &len)))
                return xstrdup(path);
 
-       char *val = pl_env_getenv_normalized(var);
-       free(var);
+       const char *val = pl_env_cache(var, len);
        if (!val)
                return xstrdup(path);
-       if (!*val || pl_env_contains_marker(val)) {
-               free(val);
-               return xstrdup(path);
-       }
 
+       const char *rem = pl_env_var_remainder(path, len);
        size_t val_len = strlen(val);
        size_t rem_len = strlen(rem);
 
@@ -209,7 +217,6 @@ char *pl_env_restore(const char *path)
        ptr += rem_len;
        *ptr = '\0';
 
-       free(val);
        return new;
 }
 
@@ -230,17 +237,6 @@ const char *pl_env_var_remainder(const char *path, int length)
        return path+length+2;
 }
 
-char *pl_env_var_str(const char *path, const char **remainder)
-{
-       int len;
-       const char *var;
-       if (!(var = pl_env_var(path, &len)))
-               return NULL;
-       if (remainder)
-               *remainder = pl_env_var_remainder(path, len);
-       return xstrndup(var, len);
-}
-
 int pl_env_var_len(const char *path)
 {
        int len;
diff --git a/pl_env.h b/pl_env.h
index 6899ab7..74533c5 100644
--- a/pl_env.h
+++ b/pl_env.h
@@ -171,6 +171,16 @@
  */
 #define PL_ENV_DELIMITER '\x1F'
 
+/**
+ * pl_env_cache has cached and normalized environment variables. If called with
+ * NULL, the cache is regenerated from the environment. If called with a
+ * variable name, the name is trimmed and the matching environment variable is
+ * returned. The value of the environment variable is the normalized path. If
+ * var is not NULL and var_len is not -1, the first var_len characters of var
+ * are used.
+ */
+const char *pl_env_cache(const char *var, int var_len);
+
 /**
  * pl_env_save checks the base path against the configured environment
  * variables, replaces the first match with a substitution, and returns a
@@ -204,13 +214,6 @@ const char *pl_env_var(const char *path, int *out_length);
  */
 const char *pl_env_var_remainder(const char *path, int length);
 
-/**
- * pl_env_var_str is like pl_env_var, but it returns a malloc'd null-terminated
- * string instead. If remainder is not NULL, it is set to point to the remainder
- * of the path.
- */
-char *pl_env_var_str(const char *path, const char **remainder);
-
 /**
  * pl_env_var_len returns the length of the substituted environment variable
  * name, if present. Otherwise, it returns 0.
diff --git a/ui_curses.c b/ui_curses.c
index 670a00b..89db534 100644
--- a/ui_curses.c
+++ b/ui_curses.c
@@ -51,6 +51,7 @@
 #include "mixer.h"
 #include "mpris.h"
 #include "locking.h"
+#include "pl_env.h"
 #ifdef HAVE_CONFIG
 #include "config/curses.h"
 #include "config/iconv.h"
@@ -2477,6 +2478,8 @@ int main(int argc, char *argv[])
                }
        }
 
+       pl_env_cache(NULL, 0);
+
        setlocale(LC_CTYPE, "");
        setlocale(LC_COLLATE, "");
        charset = getenv("CMUS_CHARSET");
```

</details>

Reply:
I've pushed this.